docs/madr/decisions/057-sni-in-the-resource.md

```diff
@@ -0,0 +1,171 @@
+
+```yaml
+kind: MeshService
+spec:
+  ports:
+    - port: 80
+      targetPort: 8080
+      snis: # new field
+      - value: "ae10a8071b8a8eeb8.backend.8080.demo.ms"
+```
+
+The `snis` field will be filled out automatically by the original Zone CP and synced cross zone.
+The field will be filled out by the same mechanism as Mesh Defaulter, which is webhook on Kubernetes and `ResourceManager` on Universal.
+This way we can avoid a situation that we drop SNI by the resource update.
+
+**Migration**
+SNIs is a list, not a field, so we can perform a migration of SNI. If we ever change the format we can provide multiple snis in the list.
+The client should use the first SNI on the list. This way we can go from
+```yaml
+snis:
+- value: old
+```
+to
+```yaml
+snis:
+- value: new
+- value: old
+```
+to
+```yaml
+snis:
+- value: new
+```
+
+Item in an array is an object with just `value` field.
+It's an object, so we can potentially expand this field if we need it to add fields like `type` or `version` etc.
+
+Advantages:
+* Slightly better debuggability because SNI is in MeshService object
+* Easy migration
+* Easy usage from the client side
+
+Disadvantages:
+* Putting more stuff in the object
+
+#### Option 2 - Keep it implicit
+
+Advantages:
+* No new fields in objects
+
+Disadvantages:
+* Keep the existing tradeoffs
+
+### MeshMultiZoneService
+
+SNI should always be filled out by the source of truth of the resource.
+In case of `MeshMultiZoneService`, this is global CP.
+However, the problem is that `MeshMultiZoneService` does not require defining ports directly.
+They are derived from the `MeshServices` selected by `MeshMultiZoneService`.
+Global CP does not have list of ports, so it cannot update them with SNI.
+
+```yaml
+kind: MeshMultiZoneService
+name: test-server
+spec:
+  selector:
+    meshService:
+      matchLabels:
+        kuma.io/display-name: test-server
+        k8s.kuma.io/namespace: mzmsconnectivity
+status: # computed on the zone
+  vips:
+    - ip: 241.0.0.1
+  meshServices:
+    - name: test-server.mzmsconnectivity
+    - name: test-server-234jhg34yt34.kuma-system
+  ports:
+    - port: 80
+      targetPort: 8080
+```
+
+#### Option 1 - SNI in the resource definition and require ports from user
+
+We can require ports from user, so it's
+
+```yaml
+spec:
+  selector:
+    meshService:
+      matchLabels:
+        kuma.io/display-name: test-server
+        k8s.kuma.io/namespace: mzmsconnectivity
+  ports:
+    - port: 80
+      targetPort: 8080
+      snis:  # new field
+      - value: "ae10a8071b8a8eeb8.backend.80.demo.mzms"
+```
+
+SNI is then applied the same way as MeshService (via webhook/ResourceManager) and is synced down to zones.
+
+Advantage:
+* We see ports in the GUI
+* We avoid mistakes when ports are different between services. What to do when
+  * One service has port 80 and other has port 8080
+  * One service has appProtocol http on port 80 and other has tcp on port 80
```

Comment:
MeshMZService is similar to MeshService. This is something that confused @johncowen today